Feature/compound reward

codespool opened the pull request:
**Pull Request Summary**

> Adds automatic restaking of reward to `claim_staker` extrinsic by checking the `ledger.reward_handling` property and staking the claimed amount right after claim.

**Check list**
- [ ] contains breaking changes
- [X] adds new feature
- [X] modifies existing feature (bug fix or improvements)
- [ ] relies on other tasks
- [ ] documentation changes
- [X] tests and/or benchmarks are included
- [ ] changed API client type definition or chain metadata

**This pull request makes the following changes:**

**Adds**
- `ledger.reward_handling` check to `claim_staker`
- `RewardAndRestake`, `RewardHandlingChange` events
- `enable_compound_staking` extrinsic
- `update_staking_values` helper function
- tests to check claiming after unstake or manually switching `reward_handling`

**Fixes**
- fix `assert_claim_staker` test utility function to take into account `reward_handling` state 

**Changes**
- `claim_staker` now checks for `reward_handling` state, and restakes the claimed value if needed
- `bond_and_stake` uses `update_staking_values` helper function

**To-dos**
- [ ] update docs

codespool commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -890,6 +862,63 @@ pub mod pallet {
+                Error::<T>::InsufficientValue,
+            );
+
+            let mut ledger = Self::ledger(&staker);
```
pass instead of reading again

hoonsubin commented:
Sorry! I didn't mean to remove the branch! I think I made a mistake when removing my local branch. Please reopen this!

Commit:
extract staking logic

Commit:
split stake method

Commit:
handle return option

Commit:
add restake event

Commit:
add check for zero staked

Commit:
add enable_compound_staking extrinsic

Commit:
pass era instead of reading again

Commit:
update weights

Commit:
narrow down shared part of claim

Commit:
fix typos

Commit:
simplify update_staking_values

Commit:
simplify enable_compound_staking extrinsic

Commit:
cleanup

Commit:
fix assert_claim_staker

Commit:
add RewardHandlingChange event

Commit:
add enable_compound_staking to benchmark

Commit:
add test scenarios for claim only payout

Commit:
fix typos

Commit:
fix runtime-benchmarks build

Commit:
fix after rebase

Dinonard commented on frame/dapps-staking/src/benchmarking.rs:
```diff
@@ -230,6 +230,16 @@ benchmarks! {
+    verify {
+        assert_last_event::<T>(Event::<T>::RewardHandlingChange(staker, option).into());
+    }
+
```
Good!

You will also need to create separate `claim_staker` benchmarking for both cases: using compounding and not using it.

You can find examples in existing substrate pallets.

Dinonard commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -864,6 +884,22 @@ pub mod pallet {
         }
+
+        #[pallet::weight(T::WeightInfo::enable_compound_staking())]
+        pub fn enable_compound_staking(
```
Perhaps `set_reward_destination` is more general?

Btw, enum has been renamed to `RewardDestination` in latest PR version.

Dinonard commented on frame/dapps-staking/src/testing_utils.rs:
```diff
@@ -419,13 +419,23 @@ pub(crate) fn assert_claim_staker(claimer: AccountId, contract_id: &MockSmartCon
+            claim_era,
+            calculated_reward,
+        )));
+    }
```
I believe you should also expand the checked storage values AFTER the call.

E.g. locked/staked values should be increased for 3 separate storage entries.
In case compounding isn't used, they should remain the same.

Dinonard commented on frame/dapps-staking/src/testing_utils.rs:
```diff
@@ -419,13 +419,23 @@ pub(crate) fn assert_claim_staker(claimer: AccountId, contract_id: &MockSmartCon
-    )));
 
+    if init_state.ledger.reward_handling == RewardHandling::PayoutAndStake
+        && !init_state.staker_info.latest_staked_value().is_zero()
```
Can this be a helper method?

Dinonard commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -872,6 +908,25 @@ pub mod pallet {
             T::PalletId::get().into_account()
         }
 
+        fn update_staking_values(
```
This is just my opinion, but I'm not very fond of this utility method.

It does too much different things, imo.
Updates 3 distinct values, but only one gets written into storage.
I understand that the purpose is to avoid code duplication but perhaps it's fine to have these few lines duplicated.

What do you think?

codespool replied:
agreed, I imagined it differently when I started, but now it's just proxying a call to two mutations.
removed.

Dinonard commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -741,15 +735,41 @@ pub mod pallet {
+            let mut ledger = Self::ledger(&staker);
 
+            if ledger.reward_handling == RewardHandling::PayoutAndStake
+                && !staker_info.latest_staked_value().is_zero()
```
Come to think of it, there's also a third-scenario: contract has been unregistered.

Perhaps this should be a utility method?

Dinonard commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -741,15 +735,41 @@ pub mod pallet {
+            if ledger.reward_handling == RewardHandling::PayoutAndStake
+                && !staker_info.latest_staked_value().is_zero()
+            {
+                staker_info
+                    .stake(current_era, staker_reward)
+                    .map_err(|_| Error::<T>::UnexpectedStakeInfoEra)?;
```
I've been thinking about this and believe we have to handle scenario when number of `EraStake` values in `stakes` vector is exceeded.

### Example
// <era, staked amount>
**stakers** = [`<1, 100>`, `<95, 300>`]

This tells us that staker had staked **100** in era **1** and then increased his stake to **300** in era **95**.
NO era rewards have been claimed.

No consider a scenario where staker decides to do the following:
* Claim in era **96** with compounding rewards. **stakers** becomes [<...>, <...>, `<96, 301>`]
* Claim in era **97** with compounding rewards. **stakers** becomes [<...>, <...>, `<96, 301>`, `<97, 302>`]
* ...

By doing this repeatedly, vec can grow to uncontrolled size.

### Solution
* Add a check for `stakers` vector size.
* I will modify the original PR to allow `bond_and_stake` and `unbond_and_unstake` to only allow vec to grow up to `MaxEraStakeValues - 1`.
* This will allow batch call of `claim_staker` to claim multiple eras at once.
* If staker gets stuck, they can always switch **compounding rewards** off and reduce the size of `stakers` vec.

What do you think about that?

codespool replied:
Makes sense. I added the check.

Dinonard commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -741,15 +735,41 @@ pub mod pallet {
+                    contract_id.clone(),
+                    era,
+                    staker_reward,
+                ));
```
It is important we still emit the old event as before.

External services use it to calculate how much rewards were earned by a particular account.
If we change this, they all need to update too and we want to avoid that.

You can just emit two separate events.

Dinonard commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -723,7 +717,7 @@ pub mod pallet {
 
-            let staking_info = Self::staking_info(&contract_id, era);
+            let mut staking_info = Self::staking_info(&contract_id, era);
             let reward_and_stake =
```
You already read `EraInfo` here so you shouldn't do it again in the utility method `update_staking_values`

codespool replied:
removed the utility method.

Dinonard commented on frame/dapps-staking/src/tests.rs:
```diff
@@ -1289,6 +1289,51 @@ fn claim_after_unregister_is_ok() {
+        assert_bond_and_stake(staker, &contract_id, stake_value);
+
+        advance_to_era(start_era + 1);
+        assert_ok!(DappsStaking::enable_compound_staking(
+            Origin::signed(staker),
+            RewardHandling::OnlyPayout
+        ));
```
I suggest you add an `assert_` utility function for this where you assert the event and init/final state.

Dinonard commented on frame/dapps-staking/src/tests.rs:
```diff
@@ -1289,6 +1289,51 @@ fn claim_after_unregister_is_ok() {
+        let stake_value = 100;
+        assert_bond_and_stake(staker, &contract_id, stake_value);
+
+        advance_to_era(start_era + 1);
```
Perhaps a few inline comments describing what's being done would be helpful?

Dinonard commented on frame/dapps-staking/src/tests.rs:
```diff
@@ -1289,6 +1289,51 @@ fn claim_after_unregister_is_ok() {
+
+        assert_claim_staker(staker, &contract_id);
+    })
+}
```
How about covering the new extrinsic call with UT?

Also, these scenarios should be covered:
* switch between different reward payouts withing one UT and verify it works
* Produce a scenario where `stakers` vec size would go out of limit
  * verify that if user switches off compounding rewards, claim can once again work

Dinonard reviewed (commented):
Good progress! :+1:

codespool committed:
cleanup after rebase

codespool committed:
pr fix: change reward handling to destination

codespool committed:
remove update_staking_info method

codespool committed:
fix merge artefact

codespool committed:
fix merging bugs

codespool committed:
keep the Reward event in case of FreeBalance dest.

codespool committed:
add check for stakes length

codespool committed:
extract restake check to util method

Dinonard commented:
Please rebase on top of latesat `polkadot-v0.9.17` and I'll check the code.

codespool committed:
tests: add check for both claim events

codespool committed:
tests: use helper method to detect restaking

codespool committed:
tests: add memory snapshots for current era

codespool committed:
tests: add assertion for set_reward_destination

codespool committed:
tests: add more coverage

- claim_with_different_reward_destination_is_ok
- claiming_when_stakes_full_without_compounding_is_ok

codespool committed:
fix: move withdraw mutation after stakes len check

codespool committed:
benchmark and update weights

codespool committed:
fix: update legacy weights

Dinonard commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -253,6 +253,10 @@ pub mod pallet {
         /// Maintenance mode has been enabled or disabled
         MaintenanceMode(bool),
+        /// Claimed staker reward restaked
+        RewardAndRestake(T::AccountId, T::SmartContract, BalanceOf<T>),
```
Should we just reuse `BondAndStake` event?

After all, it does convey the same information.

codespool replied:
Yes, makes sense.

Dinonard commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -582,7 +580,7 @@ pub mod pallet {
             Self::update_ledger(&staker, ledger);
             Self::update_staker_info(&staker, &contract_id, staker_info);
-            ContractEraStake::<T>::insert(&contract_id, current_era, staking_info);
+            ContractEraStake::<T>::insert(contract_id.clone(), current_era, staking_info);
```
Is this change needed?
Why can't reference be used?

codespool replied:
Good catch, thanks!

Dinonard commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -751,6 +750,44 @@ pub mod pallet {
+
+            if Self::should_restake_reward(
+                ledger.reward_destination,
+                &contract_id,
```
You shouldn't read `DAppInfo` twice.

It's already read on L734.

Dinonard commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -751,6 +750,44 @@ pub mod pallet {
+                staker_info.latest_staked_value(),
+            ) {
+                // There must be one slot left for restaking
+                ensure!(
+                    staker_info.len() < T::MaxEraStakeValues::get(),
+                    Error::<T>::TooManyEraStakeValues
+                );
```
This isn't correct - you should do the check with `<=` AFTER calling `stake` on `staker_info` struct.

This is important since in a scenario like this:
```
current_era = 15;
MaxEraStakeValues = 3;
stakes = [<1000, 5>, <2000, 10>]

UI prepares a batch call of claims (e.g. let's say there are 10 claim_staker calls).
1. [<1000, 6>, <2000, 10>, <2001, 15>]
2. ERROR
3. ERROR
...
10. ERROR
```
But it's not a problem since an existing `EraStake` value will be modified.

Please also add UT for this scenario.

codespool replied:
in case of a batch call, only the last item would be mutated, right?
so, something like:
```rust
ensure!(
    staker_info.len() < T::MaxEraStakeValues::get() 
    || staker_info.latest_staked_era == current_era,
    Error::<T>::TooManyEraStakeValues
);
```
should make the result:
```
1. [<1000, 6>, <2000, 10>, <2001, 15>] 
2. [<1000, 7>, <2000, 10>, <2002, 15>] 
3. [<1000, 8>, <2000, 10>, <2003, 15>]
4. [<1000, 9>, <2000, 10>, <2004, 15>]
5. [<2000, 10>, <2005, 15>]
6. [<2000, 11>, <2007, 15>]
7. [<2000, 12>, <2009, 15>]
8. [<2000, 13>, <2011, 15>]
9.[<2000,14>,<2013,15>]
10.[<2015, 15>]
```

Should I add the `last_staked_era` function and check for it here?

Dinonard replied:
Yes, that's correct.
I gave this batch example just for the demonstration of how it should work.
You don't need to use `last_staked_era` for this check. :)

Just use the exact same check as in `bond_and_stake` and `unbond_and_unstake`.
With a small difference: use `<=` instead of `<`.

Just move L761 code below L766.

codespool replied:
I extracted the condition and moved it before the call to `T::Currency::withdraw` because the body of the `if` clause also had possible errors.

Dinonard commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -751,6 +750,44 @@ pub mod pallet {
+                    contract_id.clone(),
+                    staker_reward,
+                ));
+            }
```
You cannot do this here since you have code which can possibly fail on L792.

You should probably move that part BEFORE this `if` clause.

Dinonard commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -760,16 +797,19 @@ pub mod pallet {
-
-            Ok(().into())
+            Self::update_staker_info(&staker, &contract_id, staker_info);
+            Ok(Some(
+                T::WeightInfo::claim_staker_with_restake()
+                    .max(T::WeightInfo::claim_staker_without_restake()),
```
At this point you should return the correct one which was used.

codespool replied:
I wasn't clear on how to do it from the examples. Should I just `if` the return value?

Dinonard replied:
Sure!

I guess if you store the `bool` of from L755 check, it becomes a simple `if` :)

Dinonard commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -760,16 +797,19 @@ pub mod pallet {
             ));
-
-            Ok(().into())
+            Self::update_staker_info(&staker, &contract_id, staker_info);
```
Not sure I understand why the move is needed - can you elaborate?

codespool replied:
it's not if I call update_staker_info first :+1:

Dinonard commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -253,6 +253,10 @@ pub mod pallet {
+        /// Claimed staker reward restaked
+        RewardAndRestake(T::AccountId, T::SmartContract, BalanceOf<T>),
+        /// Reward handling modified
+        RewardDestinationSet(T::AccountId, RewardDestination),
```
Can we use just `RewardDestination`?

Imo, it's more concise and conveys the same info.

codespool replied:
Sure we can.. get and set sufixes are just old habits :D

Dinonard commented on frame/dapps-staking/src/tests.rs:
```diff
@@ -1360,6 +1360,123 @@ fn claim_after_unregister_is_ok() {
         }
     })
 }
+#[test]
```
Please separate it from above UT with new line.

Dinonard commented on frame/dapps-staking/src/tests.rs:
```diff
@@ -1360,6 +1360,123 @@ fn claim_after_unregister_is_ok() {
+        // it's filled by claiming and restaking once
+        assert_claim_staker(staker_id, &contract_id);
+
+        // claiming should not work because it can't stake any more
+        assert_noop!(
+            DappsStaking::claim_staker(Origin::signed(staker_id), contract_id),
+            Error::<TestRuntime>::TooManyEraStakeValues
+        );
```
It should work :)

codespool replied:
Should it? The call before has filled the `stakes` vector, and it shouldn't be possible to claim with reward_destination set to `stakeBalance`, which is changed to `freeBalance` in the next call.

Dinonard replied:
I think this is related to: https://github.com/AstarNetwork/astar-frame/pull/14#discussion_r830860865

Since era remains the same, it's not possible for consecutive calls of `claim_staker` to increase the size of the vector.

Dinonard commented on frame/dapps-staking/src/testing_utils.rs:
```diff
@@ -394,43 +394,93 @@ pub(crate) fn assert_withdraw_unbonded(staker: AccountId) {
+    let final_state_claim_era = MemorySnapshot::all(claim_era, contract_id, claimer);
+    let final_state_current_era = MemorySnapshot::all(current_era, contract_id, claimer);
+
+    // If there's less than 2 events, it should fail anyway, so panic as acceptable
+    let events = dapps_staking_events();
+    let second_last_event = &events[events.len() - 2];
```
I'd recommend you clear all events prior to calling `claim_staker` to ensure you don't have any leftovers from old calls.

Dinonard commented on frame/dapps-staking/src/testing_utils.rs:
```diff
@@ -394,43 +394,93 @@ pub(crate) fn assert_withdraw_unbonded(staker: AccountId) {
+    ) {
+        assert_eq!(
+            second_last_event.clone(),
+            Event::<TestRuntime>::RewardAndRestake(claimer, contract_id.clone(), calculated_reward,)
```
You're only checking whether one event is correct in this case.

codespool replied:
yes, the last event is asserted regardless of the `if` clause (`Reward` should be dispatched in both cases)

Dinonard commented on frame/dapps-staking/src/testing_utils.rs:
```diff
@@ -394,43 +394,93 @@ pub(crate) fn assert_withdraw_unbonded(staker: AccountId) {
+            final_state_current_era.era_info.locked
+        );
+    } else {
+        assert_eq!(
+            init_state_claim_era.free_balance + calculated_reward,
+            final_state_claim_era.free_balance
+        );
```
Isn't this check also applicable if `if` clause is `true`?

codespool replied:
yes, it's a mistake, it should check free balance in the `current_era` is higher after claim without compounding

Dinonard commented on frame/dapps-staking/src/testing_utils.rs:
```diff
@@ -394,43 +394,93 @@ pub(crate) fn assert_withdraw_unbonded(staker: AccountId) {
+    let events = dapps_staking_events();
+    let second_last_event = &events[events.len() - 2];
+
+    if DappsStaking::should_restake_reward(
```
Since this is `pub`, you should also probably add a few UTs for it.

Also, this function has become quite large - maybe having separate helper functions for case when no restake is done and when it is would help increase readability.

Dinonard reviewed (commented):
Please add some comments to functions/methods that are missing them.

codespool committed:
pr fix: reuse BondAndStake event

codespool committed:
pr fix: rename RewardDestination event

codespool committed:
pr fix: prevent unnecessary clone and read

codespool committed:
pr fix: move mutating parts before possible error

codespool committed:
pr fix: move all checks before mutations

codespool committed:
pr fix: nicer conditional weight return

codespool committed:
pr fix: clear old events, assert correct era info

codespool committed:
pr fix: improve readability, add comments

codespool committed:
pr fix: add UT for should_restake_reward

Dinonard commented on frame/dapps-staking/src/benchmarking.rs:
```diff
@@ -195,7 +195,7 @@ benchmarks! {
     }
 
-    claim_staker {
+    claim_staker_with_restake {
```
One small change I'd recommend here: set the restake option explicity.

If by any chance we modify the default value in the future, this won't need update.

Dinonard commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -562,15 +564,9 @@ pub mod pallet {
-                .checked_add(&value_to_stake)
-                .ok_or(ArithmeticError::Overflow)?;
+            // Increment ledger and total staker value for contract.
+            ledger.locked = ledger.locked.saturating_add(value_to_stake);
+            staking_info.total = staking_info.total.saturating_add(value_to_stake);
```
Interesting enough, I've found a neat way how we can inject `T: Config` as dependency into these structs without causing us too much problems.

So basically, we could have this method return an `Result<_,_>` and just use `?` operator to either get the ok value or return error.

Not for this commit though :)

If you're interested, check `pallet-staking` in substrate repo.

codespool replied:
nice! will take a look 👍

Dinonard commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -751,6 +748,25 @@ pub mod pallet {
+                    .stake(current_era, staker_reward)
+                    .map_err(|_| Error::<T>::UnexpectedStakeInfoEra)?;
+
+                ensure!(
```
Would an inline comment here be a good idea?

Since we have a different check here and in `bond_and_stake` + `unbond_and_unstake` calls.

Dinonard commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -759,17 +775,39 @@ pub mod pallet {
-            ));
-
-            Ok(().into())
+            Ok(Some(if should_restake_reward {
+                T::WeightInfo::claim_staker_with_restake()
+            } else {
+                T::WeightInfo::claim_staker_without_restake()
+            })
+            .into())
```
Nice :)

Dinonard commented on frame/dapps-staking/src/testing_utils.rs:
```diff
@@ -394,43 +394,68 @@ pub(crate) fn assert_withdraw_unbonded(staker: AccountId) {
 
     // Cannot claim rewards post unregister era, this indicates a bug!
-    if let DAppState::Unregistered(unregistered_era) = init_state.dapp_info.state {
+    if let DAppState::Unregistered(unregistered_era) = init_state_claim_era.dapp_info.state {
```
I missed this before. Good thing you noticed it!

codespool replied:
well, yeah, it took me a while to figure out snapshots capture a specific era :)
I created two separate snapshots, one for `claim_era`, one for `current_era`, do you mind double checking the logic for the assertions now, in case I messed something up?

Dinonard replied:
I did during review :)

It looks good to me!

Dinonard commented on frame/dapps-staking/src/testing_utils.rs:
```diff
@@ -446,6 +471,41 @@ pub(crate) fn assert_claim_staker(claimer: AccountId, contract_id: &MockSmartCon
     assert_eq!(issuance_before_claim, issuance_after_claim);
 }
 
+// assert staked and locked states depending on should_restake_reward
+// returns should_restake_reward result so further checks can be made
+fn assert_restake_reward(
```
Great that we have a util methods for this.

Can we keep them separated though?
We use `assert_` style methods as non-return functions which cause a panic in case some expected condition isn't met.

I'd recommend just removing `-> bool`, and calling `should_restake_reward` explicitly where you need it.

Dinonard commented on frame/dapps-staking/src/tests.rs:
```diff
@@ -1361,6 +1361,124 @@ fn claim_after_unregister_is_ok() {
+        assert_bond_and_stake(staker, &contract_id, stake_value);
+        advance_to_era(start_era + 1);
+
+        // ensure reward_destination is set tot StakeBalance
```
You should disable spell checker German settings :grin:

codespool replied:
hahah, I think it ignores the words with <= 3 characters 😄

Dinonard commented on frame/dapps-staking/src/tests.rs:
```diff
@@ -1361,6 +1361,124 @@ fn claim_after_unregister_is_ok() {
+
+        // enable compounding mode, wait 3 eras
+        assert_set_reward_destination(staker, RewardDestination::StakeBalance);
+        advance_to_era(start_era + 3);
```
I'm not sure I understand why this part is needed - could you explain?

Could we just advance two eras, call `claim_staker`, change reward destination, call `claim_staker` again?

codespool replied:
oh, it's actually a typo (both cases advancing to `start_era + 3`, it should be +1 and +2).
Idea was to test both cases, regardless of the default `RewardDestination`.
So, 
set destination to `FreeBalance` --> advance era --> claim --> set destination to `StakeBalance` --> advance era --> claim
 Advancing eras in between claims so there's something to claim.

Dinonard commented on frame/dapps-staking/src/tests.rs:
```diff
@@ -1361,6 +1361,124 @@ fn claim_after_unregister_is_ok() {
+        // Insert a contract under registered contracts.
+        assert_register(10, &contract_id);
+
+        // Stake with MAX_NUMBER_OF_STAKERS - 1 on the same contract. It must work.
```
MAX_ERA_STAKE_VALUES typo

Dinonard commented on frame/dapps-staking/src/tests.rs:
```diff
@@ -1361,6 +1361,124 @@ fn claim_after_unregister_is_ok() {
+        assert_claim_staker(staker_id, &contract_id);
+
+        // making another gap in eras and trying to claim and restake would exceed MAX_ERA_STAKE_VALUES
+        advance_to_era(MAX_ERA_STAKE_VALUES * 5);
```
Maybe jusy `current_era() +1`? :)

The current argument looks out of place.

Dinonard commented on frame/dapps-staking/src/weights.rs:
```diff
@@ -29,7 +31,7 @@ impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
 	// Storage: DappsStaking PreApprovalIsEnabled (r:1 w:0)
 	fn register() -> Weight {
-		(32_139_000 as Weight)
+		(60_000_000 as Weight)
```
We'll have to see about all these changes...we have a problem with current weights.

I ran them on my laptop which is more powerful than the referent machine which should be used to measure weights. So currently, all of our weights are a bit too small!

But it seems your execution increased weights quite a bit - so maybe your machine is closer to the referent machine specs. Did you ensure nothing else is running when you ran benchmarks?

Anyhow, you could try benchmarking `pallet-balances` to generate your personal weights file and then compare it to the official weights in substrate repo. That should tell you in a sense how close your machine to referent machine is.

codespool replied:
I tried benchmarking `pallet_balances`, and I also get the double that of the official repo.
I'm running it on macbook pro, 2,3 GHz 8-Core Intel Core i9, 16 GB 2667 MHz DDR4, which is considerably weaker than the reference machine.

Dinonard replied:
Oh!

In that case, perhaps it's better I run benchmarks on my machine just so we keep it consistent.

Dinonard commented on frame/dapps-staking/src/weights.rs:
```diff
@@ -122,23 +139,30 @@ impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
+	// Storage: DappsStaking Ledger (r:1 w:1)
+	fn set_reward_destination() -> Weight {
+		(24_000_000 as Weight)
+			.saturating_add(T::DbWeight::get().reads(2 as Weight))
```
Can this be optimized?
Do you need 2 reads or is 1 enough?

codespool replied:
not sure what you mean, 1 read is to check is pallet disabled, the other is to get ledger for the origin account.
Can I drop one of them?

Dinonard replied:
Oh my bad, it skipped my mind that it's a struct we need to modify - not just a simple entry.

Disregard the comment.

Dinonard reviewed (commented):
Very good! :+1: 

A few small comments but otherwise I think you're done.

codespool committed:
pr fix: explicitly set reward dest. in benchmark

codespool committed:
pr fix: typos and comments

codespool committed:
pr fix: claim test case era advancement

codespool committed:
pr fix: remove assert_restake_reward bool return

codespool committed:
pr fix: use util method for disabled check

Dinonard commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -967,7 +969,7 @@ pub mod pallet {
         }
 
-        /// `Ok` if pallet disabled for maintenance, `Err` otherwise
+        /// `Err` if pallet disabled for maintenance, `Ok` otherwise
```
Good catch :see_no_evil:

Dinonard commented on frame/dapps-staking/src/testing_utils.rs:
```diff
@@ -473,17 +479,12 @@ pub(crate) fn assert_claim_staker(claimer: AccountId, contract_id: &MockSmartCon
-    reward: u128,
-) -> bool {
-    let should_restake_reward = DappsStaking::should_restake_reward(
+fn assert_restake_reward(init_state: &MemorySnapshot, final_state: &MemorySnapshot, reward: u128) {
```
Better to use `Balance` as type for reward.

Dinonard reviewed (commented):
LGTM :+1: 

This is an approval basically, we just have to do it in order as discussed on Slack.

Well done! :smile:

codespool committed:
pr fix: change reward param type to Balance

codespool committed:
rebase fixes

codespool commented:
@Dinonard rebased 👍

Dinonard commented:
@codespool let's get this merged and deployed on Shibuya :+1: 

Can you please prepare a PR in `Astar` repo which adds compounding rewards to `local` and `Shibuya`?
You can just use this branch as dependency.

Once you have that, I can run benchmarks and give you the values so you can add them into this PR.

Dinonard commented on frame/dapps-staking/src/weights.rs:
```diff
@@ -256,6 +295,12 @@ impl WeightInfo for () {
+		(24_000_000 as Weight)
+			.saturating_add(RocksDbWeight::get().reads(2 as Weight))
 			.saturating_add(RocksDbWeight::get().writes(1 as Weight))
 	}
```
```Rust
	// Storage: DappsStaking PalletDisabled (r:1 w:0)
	// Storage: DappsStaking GeneralStakerInfo (r:1 w:1)
	// Storage: DappsStaking RegisteredDapps (r:1 w:0)
	// Storage: DappsStaking CurrentEra (r:1 w:0)
	// Storage: DappsStaking ContractEraStake (r:1 w:1)
	// Storage: DappsStaking GeneralEraInfo (r:2 w:1)
	// Storage: DappsStaking Ledger (r:1 w:1)
	// Storage: Balances Locks (r:1 w:1)
	// Storage: System Account (r:1 w:1)
	fn claim_staker_with_restake() -> Weight {
		(78_506_000 as Weight)
			.saturating_add(T::DbWeight::get().reads(10 as Weight))
			.saturating_add(T::DbWeight::get().writes(6 as Weight))
	}
	// Storage: DappsStaking PalletDisabled (r:1 w:0)
	// Storage: DappsStaking GeneralStakerInfo (r:1 w:1)
	// Storage: DappsStaking RegisteredDapps (r:1 w:0)
	// Storage: DappsStaking CurrentEra (r:1 w:0)
	// Storage: DappsStaking ContractEraStake (r:1 w:0)
	// Storage: DappsStaking GeneralEraInfo (r:1 w:0)
	// Storage: DappsStaking Ledger (r:1 w:0)
	fn claim_staker_without_restake() -> Weight {
		(56_575_000 as Weight)
			.saturating_add(T::DbWeight::get().reads(7 as Weight))
			.saturating_add(T::DbWeight::get().writes(1 as Weight))
	}
	// Storage: DappsStaking PalletDisabled (r:1 w:0)
	// Storage: DappsStaking Ledger (r:1 w:1)
	fn set_reward_destination() -> Weight {
		(15_489_000 as Weight)
			.saturating_add(T::DbWeight::get().reads(2 as Weight))
			.saturating_add(T::DbWeight::get().writes(1 as Weight))
	}
```

codespool replied:
thnx!

Dinonard commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -845,7 +885,24 @@ pub mod pallet {
+            Self::ensure_pallet_enabled()?;
+            let staker = ensure_signed(origin)?;
+
+            Ledger::<T>::mutate(&staker, |ledger| {
+                ledger.reward_destination = reward_destination;
```
Just noticed that this will leave a DB entry even if an account isn't staking.
This shouldn't be the case since we cleanup empty ledgers.

You should ensure that the ledger isn't empty before mutating it.
Add a UT for the scenario too.

codespool replied:
good catch 👍

codespool committed:
pr fix: check for empty ledger on set_reward_dest

Dinonard committed:
Compound reward precompile (#44)

* Compound reward precompile

* Single method for changing rd

* Update DappsStaking.sol

Fix typo

codespool committed:
pr fix: proper benchmark initialize

codespool committed:
pr fix: update weights

codespool committed:
bump versions

codespool committed:
Merge branch 'polkadot-v0.9.17' into feature/compound-reward

Dinonard commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -898,6 +900,11 @@ pub mod pallet {
             Self::ensure_pallet_enabled()?;
             let staker = ensure_signed(origin)?;
 
+            ensure!(
+                Ledger::<T>::contains_key(&staker),
+                Error::<T>::NotActiveStaker
+            );
```
We should keep it to 1 read and 1 write - `contains_key` counts as a read now.

I'd suggest to read the value from the DB, check it using the `is_empty()` method and then insert the modified value.

Dinonard commented on frame/dapps-staking/src/weights.rs:
```diff
@@ -29,7 +31,7 @@ impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
 	// Storage: DappsStaking PreApprovalIsEnabled (r:1 w:0)
 	fn register() -> Weight {
-		(32_139_000 as Weight)
+		(60_000_000 as Weight)
```
Please only modify the 3 benchmarks you impacted, let's leave the other numbers as they are for now.

codespool committed:
pr fix: modify only affected weights

codespool committed:
Merge branch 'feature/compound-reward' of github.com:AstarNetwork/astar-frame into feature/compound-reward

codespool committed:
pr fix: remove unnecessary read

Dinonard commented:
@codespool please fix the formatting and let's merge.

codespool committed:
pr fix: formatting
